server/src/main/java/org/elasticsearch/index/reindex/ReindexIndexClient.java

```diff
@@ -124,14 +126,18 @@ public void onFailure(Exception e) {
         }
     }
 
-    // TODO: Potentially add compare-and-set semantics to ensure that another node has not been assigned
-    //  this task
-    public void updateReindexTaskDoc(String taskId, ReindexTaskIndexState reindexState, ActionListener<Void> listener) {
-        index(taskId, reindexState, DocWriteRequest.OpType.INDEX, listener);
+    public void updateReindexTaskDoc(String taskId, ReindexTaskIndexState reindexState, long previousTerm, long previousSeqNo,
+                                     ActionListener<Void> listener) {
+        index(taskId, reindexState, DocWriteRequest.OpType.INDEX, true, previousTerm, previousSeqNo, listener);
     }
 
-    private void index(String taskId, ReindexTaskIndexState reindexState, DocWriteRequest.OpType opType, ActionListener<Void> listener) {
+    private void index(String taskId, ReindexTaskIndexState reindexState, DocWriteRequest.OpType opType, boolean conditional,
+                       long previousTerm, long previousSeqNo, ActionListener<Void> listener) {
         IndexRequest indexRequest = new IndexRequest(REINDEX_INDEX).id(taskId).opType(opType);
+        if (conditional) {
+            indexRequest.setIfPrimaryTerm(previousTerm);
+            indexRequest.setIfSeqNo(previousSeqNo);
```

Comment:
I think we will need specialized error handling to ensure that if a version conflict occurs here, we will simply stand down and not try to fail the job?

Reply:
Yes. We will need some way to halt the in-process reindex work.